crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -598,13 +601,36 @@ impl<'db> TypeInferenceBuilder<'db> {
         } = with_statement;
 
         for item in items {
-            self.infer_expression(&item.context_expr);
-            self.infer_optional_expression(item.optional_vars.as_deref());
+            if item.optional_vars.is_some() {
+                self.infer_definition(item);
+            } else {
+                self.infer_expression(&item.context_expr);
+            }
         }
 
         self.infer_body(body);
     }
 
+    fn infer_with_item_definition(
+        &mut self,
+        with_item: &ast::WithItem,
+        definition: Definition<'db>,
+    ) {
+        let ast::WithItem {
+            range: _,
+            context_expr,
+            optional_vars,
+        } = with_item;
+
+        let Some(optional_vars) = optional_vars.as_deref() else {
+            unreachable!("With item definition without optional vars");
+        };
+        self.infer_expression(context_expr);
+        let with_item_ty = self.infer_expression(optional_vars);
+
+        self.types.definitions.insert(definition, with_item_ty);
```

Comment:
Let's add a TODO for correct type inference here (but it's out of scope for this PR, since it requires features we don't have, like inferring types from a call).

The correct type for optional-vars here is the return type of the `__enter__` method on the type of `context_expr`.

This gets even a bit more complex when we consider unpacking, which I'll address in another comment.